File: backend/app/Models/Membership.php
<?php

namespace App\Models;

use App\Tenancy\BelongsToTenant;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\MorphMany;

/**
 * A purchased membership. Both limits apply at once: a 30 session pass that
 * also runs out after 30 days expires on whichever comes first.
 */
class Membership extends Model
{
    use BelongsToTenant, HasFactory;

    public const ACTIVE = 'active';

    public const EXPIRED = 'expired';

    public const FROZEN = 'frozen';

    public const CANCELLED = 'cancelled';

    protected $fillable = [
        'tenant_id', 'member_id', 'membership_plan_id', 'type', 'starts_at', 'ends_at',
        'total_sessions', 'remaining_sessions', 'price', 'discount', 'paid_amount',
        'status', 'frozen_from', 'frozen_until', 'created_by', 'notes',
    ];

    protected $attributes = [
        'status' => self::ACTIVE,
    ];

    protected function casts(): array
    {
        return [
            'starts_at' => 'date',
            'ends_at' => 'date',
            'frozen_from' => 'date',
            'frozen_until' => 'date',
            'price' => 'decimal:2',
            'discount' => 'decimal:2',
            'paid_amount' => 'decimal:2',
        ];
    }

    public function member(): BelongsTo
    {
        return $this->belongsTo(Member::class);
    }

    public function plan(): BelongsTo
    {
        return $this->belongsTo(MembershipPlan::class, 'membership_plan_id');
    }

    public function attendances(): HasMany
    {
        return $this->hasMany(Attendance::class);
    }

    public function invoiceItems(): MorphMany
    {
        return $this->morphMany(InvoiceItem::class, 'itemable');
    }

    public function isExpiredByDate(): bool
    {
        return $this->ends_at !== null && $this->ends_at->isBefore(today());
    }

    public function isOutOfSessions(): bool
    {
        return $this->remaining_sessions !== null && $this->remaining_sessions <= 0;
    }

    public function isUsable(): bool
    {
        return $this->status === self::ACTIVE
            && ! $this->isExpiredByDate()
            && ! $this->isOutOfSessions()
            && ! $this->starts_at->isAfter(today());
    }

    /** Days left, null for open ended memberships. */
    public function daysRemaining(): ?int
    {
        if ($this->ends_at === null) {
            return null;
        }

        return max(0, today()->diffInDays($this->ends_at, false));
    }

    public function scopeActive(Builder $query): Builder
    {
        return $query->where('status', self::ACTIVE);
    }

    public function scopeExpiringWithin(Builder $query, int $days): Builder
    {
        return $query->where('status', self::ACTIVE)
            ->whereNotNull('ends_at')
            ->whereBetween('ends_at', [today(), today()->addDays($days)]);
    }
}

File: mobile/gymflow_core/lib/src/models/membership.dart
/// A purchased membership. Both limits apply at once: a session pass with a
/// deadline expires on whichever runs out first.
class Membership {
  const Membership({
    required this.id,
    required this.type,
    required this.status,
    this.planName,
    this.startsAt,
    this.endsAt,
    this.totalSessions,
    this.remainingSessions,
    this.price = 0,
  });

  factory Membership.fromJson(Map<String, dynamic> json, {String locale = 'fa'}) {
    final plan = json['plan'] as Map?;
    final names = (plan?['name'] as Map?)?.cast<String, dynamic>();

    return Membership(
      id: json['id'] as int,
      type: json['type'] as String? ?? 'duration',
      status: json['status'] as String? ?? 'active',
      planName: names?[locale] as String? ?? names?['en'] as String?,
      startsAt: json['starts_at'] == null ? null : DateTime.tryParse('${json['starts_at']}'),
      endsAt: json['ends_at'] == null ? null : DateTime.tryParse('${json['ends_at']}'),
      totalSessions: (json['total_sessions'] as num?)?.toInt(),
      remainingSessions: (json['remaining_sessions'] as num?)?.toInt(),
      price: double.tryParse('${json['price']}') ?? 0,
    );
  }

  final int id;
  final String type;
  final String status;
  final String? planName;
  final DateTime? startsAt;
  final DateTime? endsAt;
  final int? totalSessions;
  final int? remainingSessions;
  final double price;

  bool get isActive => status == 'active';

  int? get daysRemaining {
    if (endsAt == null) return null;

    final today = DateTime.now();
    final days = endsAt!.difference(DateTime(today.year, today.month, today.day)).inDays;

    return days < 0 ? 0 : days;
  }

  /// How much of the pass is left, for the ring on the member's home screen.
  double get progress {
    if (totalSessions != null && totalSessions! > 0) {
      return (remainingSessions ?? 0) / totalSessions!;
    }

    if (startsAt == null || endsAt == null) return 1;

    final total = endsAt!.difference(startsAt!).inDays;
    if (total <= 0) return 0;

    return ((daysRemaining ?? 0) / total).clamp(0, 1).toDouble();
  }
}
